### PR TITLE
ci: install kernel from cache job.

### PR DESCRIPTION
**Pull request by jcvenegas**
Kernel cache job will try to keep and up-to-date,
if the kernel is in the job is the latest we dont build it.

Fixes: #1218

### CODE REVIEW

**Comment by chavafg** [comment]
firecracker CI failing:
```
15:06:24 Logging kata-env information:
15:06:24 /usr/share/defaults/kata-containers/configuration.toml: file /usr/share/kata-containers/vmlinux.container does not exist
15:06:25 Build step 'Execute shell' marked build as failure
```

File: .ci/install_kata_kernel.sh [hunk 1]
```diff
@@ -18,6 +18,9 @@ cidir=$(dirname "$0")
 source "${cidir}/lib.sh"
 source "/etc/os-release" || source "/usr/lib/os-release"
 
+latest_build_url="http://jenkins.katacontainers.io/job/kernel-nightly-$(uname -m)/lastSuccessfulBuild/artifact/artifacts"
```
Comment (grahamwhaley) [inline .ci/install_kata_kernel.sh]: I don't know if [kata-arch.sh](https://github.com/kata-containers/tests/blob/master/.ci/kata-arch.sh) would be a better thing than `uname` here?

File: .ci/install_kata_kernel.sh [hunk 2]
```diff
@@ -73,20 +76,14 @@ build_and_install_kernel() {
...
-	fi
-
-	return "$rc"
+install_prebuilt_kernel(){
+	info "Install pre-built kernel version"
+	local kernel_version=$(curl -qL "${latest_build_url}/latest")
+	mkdir -p "${kernel_dir}"
+	local kernel_binary_name="vmlinuz-${kernel_version}"
+	local kernel_binary_path="${kernel_dir}/${kernel_binary_name}"
+	sudo -E curl -L "${latest_build_url}/${kernel_binary_name}" -o "${kernel_binary_path}"
```
Comment (grahamwhaley) [inline .ci/install_kata_kernel.sh]: maybe wrt the fc CI fail - do we need to build and pull down the vmlinux as well as the vmlinuz?

File: .ci/install_kata_kernel.sh (hunk 1 above)
Comment (jcvenegas) [reply .ci/install_kata_kernel.sh]: I dont want to depend a lot of the script, I think uname is fine for most of the uses, I think kata-arch is great for get right paths to golang. @jodh-intel  any preference ?

Comment (jodh-intel) [reply .ci/install_kata_kernel.sh]: Well, I think we were hoping that `kata-arch.sh` would be the "Mother of all arch commands to end all arch commands" but if it doesn't buy you anything I guess stick with `uname`.

File: .ci/ci_cache_kernel.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# Copyright (c) 2019 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o errtrace
+
+function handle_error {
```
Comment (jodh-intel) [inline .ci/ci_cache_kernel.sh]: There seems to be quite a bit of duplication between this PR and #1215. Could you maybe refactor the common bits into `.ci/lib.sh`?

File: .ci/install_kata_kernel.sh [hunk 4]
```diff
@@ -73,20 +76,24 @@ build_and_install_kernel() {
...
+	local vmlinuz_binary_path="${kernel_dir}/${vmlinuz_binary_name}"
+	sudo -E curl -qL "${latest_build_url}/${vmlinuz_binary_name}" -o "${vmlinuz_binary_path}"
+	sudo -E ln -sf "${vmlinuz_binary_path}" "${kernel_dir}/vmlinuz.container"
+
+	local vmlinux_binary_name="vmlinux-${kernel_version}"
+	local vmlinux_binary_path="${kernel_dir}/${vmlinux_binary_name}"
+	sudo -E curl -qL "${latest_build_url}/${vmlinux_binary_name}" -o "${vmlinux_binary_path}"
+	sudo -E ln -sf "${vmlinux_binary_path}" "${kernel_dir}/vmlinux.container"
+	pushd "${kernel_dir}" > /dev/null
+	sudo curl -fsOL "${latest_build_url}/shasum256-kernel"
```
Comment (jodh-intel) [inline .ci/install_kata_kernel.sh]: - Do we need to use `sudo` with `curl`.
- If so, shouldn't it be `sudo -E`?

File: .ci/ci_cache_kernel.sh (hunk 3 above)
Comment (jcvenegas) [reply .ci/ci_cache_kernel.sh]: agree, lets merge #1215 first, I will label this with `DNM`, once is merged I'll refactor, I dont want to move both changes in the same PR.

Comment (jodh-intel) [reply .ci/ci_cache_kernel.sh]: Hi @jcvenegas - https://github.com/kata-containers/tests/pull/1215 has now landed.

**Comment by raravena80** [comment]
@jcvenegas ping, any updates?

**Comment by chavafg** [comment]
@jcvenegas still working on this? can you fix conflicts ?

**Comment by amshinde** [comment]
@jcvenegas Any updates?

**Comment by jcvenegas** [comment]
```
17:33:19 Install Kata Containers Kernel
17:33:19 Download and update github.com/kata-containers/packaging
17:33:19 /tmp/install-kata-7ixVEW6Rark /tmp/jenkins/workspace/kata-containers-tests-fedora-PR/go/src/github.com/kata-containers/tests
17:33:20 package github.com/kata-containers/packaging: no Go files in /tmp/jenkins/workspace/kata-containers-tests-fedora-PR/go/src/github.com/kata-containers/packaging
17:33:20 /tmp/jenkins/workspace/kata-containers-tests-fedora-PR/go/src/github.com/kata-containers/tests
17:33:20 INFO: current kernel : 4.19.28-41
17:33:20 INFO: cached kernel  : 4.19.28-41
17:33:20 INFO: Install pre-built kernel version
17:33:20 INFO: Installing vmlinux
17:33:20 
                                                                           0.2%
##                                                                         2.9%
################                                                          23.1%
####################################                                      50.9%
######################################################                    76.3%
######################################################################## 100.0%
17:33:21 INFO: Installing /usr/share/kata-containers/vmlinux-4.19.28-41 and symlink /usr/share/kata-containers/vmlinux.container
17:33:21 INFO: Installing vmlinuz
17:33:21 
                                                                           1.3%
########                                                                  12.1%
####################################################################      95.0%
######################################################################## 100.0%
17:33:21 INFO: Installing /usr/share/kata-containers/vmlinuz-4.19.28-41 and symlink /usr/share/kata-containers/vmlinuz.container
17:33:21 INFO: Verify download checksum
17:33:21 vmlinux-4.19.28-41: OK
17:33:21 vmlinuz-4.19.28-41: OK
```

Kernel installed in a few seconds

File: .ci/ci_cache_kernel.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+mkdir -p "${WORKSPACE}/artifacts"
+pushd "${WORKSPACE}/artifacts"
+rm -f sha256sum-kernel
+cache_build_kernel "${kata_dir}/vmlinux.container"
+cache_build_kernel "${kata_dir}/vmlinuz.container"
+ls -la "${WORKSPACE}/artifacts/"
+popd
+sync
```
Comment (grahamwhaley) [inline .ci/ci_cache_kernel.sh]: why a `sync`?

File: .ci/ci_cache_kernel.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+# Copyright (c) 2019 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o errtrace
+
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```
Comment (grahamwhaley) [inline .ci/ci_cache_kernel.sh]: I thought our general idiom for this didn't have the whole cd/pwd in it - maybe you can use a `readlink -f` like you do further down? This with the cd/pwd just looks odd ;-)

File: .ci/ci_cache_kernel.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+# Source to trap error line number
+source "${script_dir}/../lib/common.bash"
+
+WORKSPACE=${WORKSPACE:-$(pwd)}
+kata_dir="/usr/share/kata-containers/"
+
+cache_build_kernel() {
```
Comment (grahamwhaley) [inline .ci/ci_cache_kernel.sh]: maybe s/build/built/ - as the function name implied to me it was going to build the kernel, but I don't think it does does it? - it just caches the already built one?

File: .ci/install_kata_kernel.sh [hunk 8]
```diff
@@ -98,16 +101,15 @@ main() {
...
 	kata_config_version="$(get_kata_config_version)"
-	current_kernel_version="${kernel_version}.${kata_config_version}"
-	info "Current Kernel version ${current_kernel_version}"
-	info "Get packaged kernel version"
-	packaged_kernel_version=$(get_packaged_kernel_version)
-	info "Packaged Kernel version ${packaged_kernel_version}"
-	if [ "$packaged_kernel_version" == "$current_kernel_version" ] && [ "$kernel_arch" == "x86_64" ]; then
-		# If installing packaged kernel from OBS fails,
+	current_kernel_version="${kernel_version}-${kata_config_version}"
+	last_build_kernel_version=$(curl -sfL "${latest_build_url}/latest")
```
Comment (grahamwhaley) [inline .ci/install_kata_kernel.sh]: I think you do this same line again later on (line 72 in current patchset)? - this does not look like a local var - can you just re-share this value, to save an extra duplicate `curl`?

File: .ci/install_kata_kernel.sh [hunk 9]
```diff
@@ -98,16 +101,15 @@ main() {
...
+	info "current kernel : ${current_kernel_version}"
+	info "cached kernel  : ${last_build_kernel_version}"
+	if [ "$last_build_kernel_version" == "$current_kernel_version" ] && [ "$kernel_arch" == "x86_64" ]; then
+		# If installing kernel fails,
 		# then build and install it from sources.
-		if ! install_packaged_kernel;then
-			info "failed to install packaged kernel, trying to build from source"
+		if ! install_prebuilt_kernel; then
+			info "failed to install kernel, trying to build from source"
```
Comment (grahamwhaley) [inline .ci/install_kata_kernel.sh]: s/install kernel/install cached kernel/

File: .ci/ci_cache_kernel.sh [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+        sudo chown -R "${USER}:${USER}" ./
+	sha256sum "${kernel_binary}" >> "sha256sum-kernel"
+}
+
+mkdir -p "${WORKSPACE}/artifacts"
+pushd "${WORKSPACE}/artifacts"
+rm -f sha256sum-kernel
+cache_build_kernel "${kata_dir}/vmlinux.container"
+cache_build_kernel "${kata_dir}/vmlinuz.container"
```
Comment (jodh-intel) [inline .ci/ci_cache_kernel.sh]: Nit/idea: you could use a loop for these two kernel files.

File: .ci/ci_cache_kernel.sh (hunk 5 above)
Comment (jcvenegas) [reply .ci/ci_cache_kernel.sh]: The job is running in a VM, this is going to be collected by the jenkins master node, do you want to be added as a comment ?

Comment (grahamwhaley) [reply .ci/ci_cache_kernel.sh]: Ah, I see. Yeah, if you are updating then add a note pls - thanks!

Comment (jcvenegas) [reply .ci/ci_cache_kernel.sh]: done :)

**Comment by jcvenegas** [comment]
```
16:02:43 INFO: current kernel : 4.19.28-41
16:02:43 INFO: cached kernel  : none
16:02:43 INFO: Install kernel from sources
16:02:43 INFO: Clone config and patches
16:02:43 INFO: Config version: 41
16:02:43 INFO: Kernel version: 4.19.28
16:02:43 INFO: kernel path does not exist, will download kernel
```
@chavafg

**Pull request merged**